Support for crates with filesystem origin

mosteo opened the pull request:
Addresses #132 

Remaining work:
- [x] Test for bad path in index

pmderodat commented on src/alire/alire-toml_index.ads:
```diff
@@ -321,7 +321,8 @@ private
+      Pkg         : Package_Type;
       Environment : Environment_Variables;
       Releases    : out Containers.Release_Sets.Set);
    --  Generate one fully populated release per version loaded in Pkg
```
Could you complete the documentation to describe what value callers must provide for `Package_Dir`?

mosteo replied:
Done in new force-push

mosteo commented on src/alire/alire-toml_index.adb:
```diff
@@ -1657,6 +1660,30 @@ package body Alire.TOML_Index is
+                             Dirs.Full_Name
+                               (Package_Dir
+                                & GNAT.OS_Lib.Directory_Separator
+                                & Path);
```
Problem is, this fails with Path not being a single path component (e.g. '../../blah')

pmderodat replied:
Ah, you are right. We would need something as powerful as Python’s `os.path.join`, and it seems that GNATCOLL.VFS.Join (or the `/` operator) does exactly the job.

mosteo commented on src/alire/alire-toml_index.adb:
```diff
@@ -1657,6 +1660,30 @@ package body Alire.TOML_Index is
+            begin
+               if Path = "" then
+                  Error (+"Empty path given in local origin");
+               elsif Path (Path'First) = GNAT.OS_Lib.Directory_Separator then
```
Right you are. There will be a couple more of these through the codebase though...

pmderodat replied:
No doubt. :-) What do you think about adding `gnatcoll-core` as a new dependency to leverage its file name handling?

mosteo replied:
In the past I was reticent to add gnatcoll because it seemed a heavy dependency with some others in tow... I see now that it has been split into parts. Since you know it better, do you think gnatcoll-core will be no trouble in general for the build process? If so I'd say let's go for it.

pmderodat replied:
Until end of 2017, GNATCOLL was instead a monolithic repo with non-Ada dependencies, so packaging it was messy. Nowadays, `gnatcoll-core` only contains sources that only depend on GNAT itself. The only subtelty is that you have to pass the appropriate scenario variables depending on the platform (`-XGNATCOLL_OS=unix|windows`, `-XGNATCOLL_MMAP=yes|no`, …). As for now we test `alr` only on GNU/Linux, I guess we don’t need to worry too much about that. :-)

I would be fine to take care of adding this dependency and setting it up properly: would this be okay for you?

mosteo replied:
That's great, much simpler to use. Scenario variables are ok with me, I start to pull my hair when there are autoconf/make/file swapping steps involved. For this PR I will use GNAT.OS_Lib.Is_Absolute_Path, so I'd say that you add it once this one is finished.

mosteo commented:
> Great, thanks for this! In addition to these comments: would it be possible for you to write testcases, at least to exercize error conditions? The regular “working” case will be covered by most testcases after the refactoring to use filesystem origins to avoid network access in the testsuite. Of course I’m available for any help you would need to get familiar with the testsuite framework :-)

Sure, I intend to but this is only the first commit of several to come. With this, the local crates can be declared and the index loaded, but still I need to generate project files that use the local crates in place (because we do not want them to be copied in during `get`, right?)

mosteo commented:
@pmderodat Incidentally, I've made a test of the new e3.testsuite xunit output and it seems to be almost there. You can see the shippable report at https://app.shippable.com/github/alire-project/alire/runs/237/1/tests . It properly shows the count and identity of failed tests. The test output is missing. After checking my own output, I think it will be enough to remove the `<system-out />` element and store the log directly in the `<failure>` element body.

pmderodat commented:
> Sure, I intend to but this is only the first commit of several to come. With this, the local crates can be declared and the index loaded

Correct me if I’m wrong, but the first commit is already enough to test the error cases, right? I see several advantages for adding testcases as early as possible:

* these tests serve as examples for the desired behavior (which is sometimes hard to infer from just the patch);
* these demonstrate that the added code actually works as expected;
* having them early reduces the chance we forget to add them.

What do you think?

> but still I need to generate project files that use the local crates in place (because we do not want them to be copied in during `get`, right?)

Hm… this sound like a premature optimization to me. :-) Multiple sessions could refer to the same directory through their index, and in this case it would be invalid not to have local copies, as both would step on each other’s toes durings builds (they would share the same object directories, for instance). So I would rather be in favor of copying the sources.

pmderodat commented:
> @pmderodat Incidentally, I've made a test of the new e3.testsuite xunit output and it seems to be almost there. You can see the shippable report at https://app.shippable.com/github/alire-project/alire/runs/237/1/tests

Very nice, thank you for the experiment!

> The test output is missing. After checking my own output, I think it will be enough to remove the `<system-out />` element and store the log directly in the `<failure>` element body.

Got it, thanks! I have updated my patches to reflect that (https://github.com/pmderodat/e3-testsuite/commit/98e68472d4def40f19d40e210e9849832eb0d26a).

mosteo commented:
> Correct me if I’m wrong, but the first commit is already enough to test the error cases, right? I see several advantages for adding testcases as early as possible:

Yes and no. You can load such an index (in fact in the latest push I modified the basic_index to use local crates) but alr will bomb at `get` time, since that's unimplemented (two tests fail).

> * these tests serve as examples for the desired behavior (which is sometimes hard to infer from just the patch);
> * these demonstrate that the added code actually works as expected;
> * having them early reduces the chance we forget to add them.

Agree on all points; so I already changed basic_index despite the mentioned tests failing (that's next).

> Hm… this sound like a premature optimization to me. :-) Multiple sessions could refer to the same directory through their index, and in this case it would be invalid not to have local copies, as both would step on each other’s toes durings builds (they would share the same object directories, for instance). So I would rather be in favor of copying the sources.

I did not intend that as an optimization but as the desired behavior, but I had not considered the multiple aliasing problem. As a programmer using it to develop local crates, I suspect the no-copy behavior is what is wanted: you edit the crates were you put them in the first place, instead of getting multiple copies of them that you cannot edit consistently. What do you think?

I wonder if gprbuild has a simple way of doing an out-of-tree build of everything in an aggregate project. What about `--relocate-build-tree`? Having out of tree builds could be worthwhile in any case, as it's common practice.

> Got it, thanks! I have updated my patches to reflect that ([pmderodat/e3-testsuite@98e6847](https://github.com/pmderodat/e3-testsuite/commit/98e68472d4def40f19d40e210e9849832eb0d26a)).

I copied the changed file manually, yesterday. May I ask if there is a simple proper way of overriding the virtualenv e3.testsuite with your repo?

mosteo commented:
Answering to myself, it does seem that --relocate-build-tree works. I had #95 in mind since long ago; I might take a stab at it, time permitting (ouch).

pmderodat commented:
> Yes and no. You can load such an index (in fact in the latest push I modified the basic_index to use local crates) but alr will bomb at `get` time, since that's unimplemented (two tests fail).

This is correct, however to me the point of testing “error conditions” is only to test the behavior when errors occurs. So here, `get` (or `show`, or `list`…) would not get past the errors, and thus testing can be effective even though the rest is not implemented, no?

> Agree on all points; so I already changed basic_index despite the mentioned tests failing (that's next).

Ah! ^^’ Unlike the testing of error conditions, testing of the actual feature should wait for the features to be implemented, otherwise the testsuite won’t be clean. So I think the rewriting of `basic_index` (and others?) should wait for the implementation to be ready before happening.

> I did not intend that as an optimization but as the desired behavior, but I had not considered the multiple aliasing problem. As a programmer using it to develop local crates, I suspect the no-copy behavior is what is wanted: you edit the crates were you put them in the first place, instead of getting multiple copies of them that you cannot edit consistently. What do you think?

Good question! I need to think about what workflow I would expect from a package manager when modifying multiple crates at the same time.

> I copied the changed file manually, yesterday. May I ask if there is a simple proper way of overriding the virtualenv e3.testsuite with your repo?

It’s just a matter of installing e3-testsuite from there. I’m not sure how you install dependencies today, but running the following should be enough:
```python
pip install git+https://github.com/pmderodat/e3-testsuite.git
```

mosteo commented:
Now it works. I reordered the test commits so the test for erroneous index is at the beginning and the local-pointing index is at the end.

> Good question! I need to think about what workflow I would expect from a package manager when modifying multiple crates at the same time.

At the moment I've opted for copying on `get`. Since the GPR generated files point to the copied crates, I expect edition would likely be wanted on those. As long as source control folders are copied too, you can edit in all the copies and push/pull changes normally. And as you said, we avoid reusing build products.

This also makes the PR independent of #137.

pmderodat commented on testsuite/fixtures/crates/hello_1.0.0_8cac0afd/LICENSE:
```diff
@@ -0,0 +1,21 @@
+MIT License
```
This file is not required; I think we should avoid all unecessary files in testcase material, to simplify as much as possible the test environment. What do you think?

mosteo replied:
OK. I just copied the crates as-is, but I see your point.

pmderodat commented on testsuite/fixtures/crates/hello_1.0.0_8cac0afd/alire.gpr:
```diff
@@ -0,0 +1,6 @@
+aggregate project Alire is
```
Likewise, the whole alire-generated files don’t seem necessary: are you okay with removing them?

mosteo replied:
OK

pmderodat commented on testsuite/tests/workflows/get-build-run-clean/test.py:
```diff
@@ -1,10 +1,5 @@
 """
 Test a basic get-build-run workflow.
-
```
Yay, thanks!!

mosteo commented on testsuite/tests/index/origin-filesystem-bad-path/test.py:
```diff
@@ -0,0 +1,16 @@
+# Since the location reported is filesystem dependent, check only that the beginning
+# of the error is there:
+
+if not "ERROR: Local origin path is not a valid directory" in "".join(p.out):
```
I have never done anything too serious in Python so all guidance with it is welcome. I will incorporate these improvements.

mosteo commented:
I will get on this tomorrow (friday), but a question on methodology: for reviewing, is it better a new final "patch" commit addressing comments, or a rebase editing the appropriate commits?

pmderodat commented:
Thanks! Whether to amend commits or add new ones is a matter of taste. So I’ll tell you what I prefer :-) I prefer editting commits, as I think keeping the commit history clean is important. This is especially important when a commit was found to introduce issues: if possible, I think we should avoid leaving this “temporary flaw” land in master, for instance to avoid issues while bisecting later (and in general to ease analyzing with the history).

mosteo commented:
Since I'm not squashing PRs I see your points about fixing in-place. Unless oversight, everything should be fixed now.

mosteo commented:
Superseded by #142
